### PR TITLE
Added forwarding of selectors to grpc upstreams

### PR DESCRIPTION
**Pull request by a10zn8**

### CODE REVIEW

File: src/main/kotlin/io/emeraldpay/dshackle/rpc/Selectors.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+                        .setOrSelector(OrSelector.newBuilder().addAllSelectors(it))
+                }
+            } else if (selector.hasAndSelector()) {
+                return processMultiple(selector.andSelector.selectorsList) {
+                    Selector.newBuilder()
+                        .setAndSelector(AndSelector.newBuilder().addAllSelectors(it))
+                }
+            } else if (selector.hasNotSelector()) {
+                val inner = keepForwarded(selector.notSelector.selector)
+                return if (inner != null) {
```
Comment (MaxFomenkov) [inline src/main/kotlin/io/emeraldpay/dshackle/rpc/Selectors.kt:32]: not a kotlin way. use `inner?.let {....}`

File: src/main/kotlin/io/emeraldpay/dshackle/rpc/Selectors.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            return creator(newSelectors).setShouldBeForwarded(true).build()
+        } else if (newSelectors.size == 1) {
+            newSelectors.first()
+        } else {
+            null
+        }
+    }
+
+    fun keepForwarded(selector: Selector): Selector? {
+        if (selector.shouldBeForwarded) {
```
Comment (MaxFomenkov) [inline src/main/kotlin/io/emeraldpay/dshackle/rpc/Selectors.kt:19]: inverse IF and return null here

File: src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/JsonRpcGrpcClient.kt [hunk 3]
```diff
@@ -56,10 +54,8 @@ class JsonRpcGrpcClient(
             val req = BlockchainOuterClass.NativeCallRequest.newBuilder()
                 .setChainValue(chain.id)
 
-            if (matcher != Selector.empty) {
-                Selector.extractLabels(matcher)?.asProto().let {
-                    req.setSelector(it)
-                }
+            if (key.selector != null) {
```
Comment (MaxFomenkov) [inline src/main/kotlin/io/emeraldpay/dshackle/upstream/rpcclient/JsonRpcGrpcClient.kt:57]: KOTLIN

**Review by MaxFomenkov — CHANGES_REQUESTED** [review]

File: src/main/kotlin/io/emeraldpay/dshackle/rpc/NativeCall.kt [hunk 4]
```diff
@@ -242,6 +242,12 @@ open class NativeCall(
             val requestDecorator = getRequestDecorator(requestItem.method)
             val resultDecorator = getResultDecorator(requestItem.method)
 
+            val selector = if (request.hasSelector()) {
+                Selectors.keepForwarded(request.selector)
```
Comment (MaxFomenkov) [inline src/main/kotlin/io/emeraldpay/dshackle/rpc/NativeCall.kt:246]: `val selector =request.takeIf {it.hasSelector()}?.let{Selectors.keepForwarded(it.selector)}`

Comment (MaxFomenkov) [reply src/main/kotlin/io/emeraldpay/dshackle/rpc/NativeCall.kt:246]: if you have else { return null } in your code - it's java style, not a kotlin

**Review by MaxFomenkov — APPROVED** [review]

**Review by Termina1 — APPROVED** [review]

**Pull request merged**